Make parallel vector derive from VectorSpaceVector

kronbichler opened the pull request:
This is the first step to move `parallel::distributed::Vector` to `VectorSpaceVector`. Relates to #2535 and #745.

I now moved all code of the parallel vector to a class `LinearAlgebra::parallel::Vector`. @Rombur is this the right scheme to name things? I noticed that the inner `parallel` namespace is a bit confusing and I had to qualify some things, like the member of `ReadWriteVector`. Should we rather make it `LinearAlgebra::distributed::Vector` or `LinearAlgebra::MPI::Vector`?

Regarding the methods, I removed a few of the more obscure `operator==`, `equ` and `sadd` methods that we had in the class, whereas I put some others that are necessary when using this vector throughout the library still in the vector.

Another thing is that the vector still has access functionality as it was before (to keep the class synchronized with the rest of the library) but at some point at least parts of it could go away.

Finally, I did not put the import function in `ReadWriteVector` yet but that is easy to add.

The directly related tests are passing, the full test suite will run later today.

Rombur commented:
@kronbichler You forgot to commit ```la_parallel_vector.h``` and ```la_parallel_vectors.templates.h```

I prefer ```LinearAlgebra::distributed::Vector``` instead of ```LinearAlgebra::parallel::Vector``` It makes it easier to understand what the vector does exactly.

kronbichler commented:
@Rombur thanks, I forgot those files. They are now appended.

So right now `LinearAlgebra::distributed::Vector` is the favorite name for it. We can wait a bit to see what others think.

Rombur commented on include/deal.II/base/partitioner.h:
```diff
@@ -58,7 +59,7 @@ namespace Utilities
      * @author Katharina Kormann, Martin Kronbichler, 2010, 2011
      */
-    class Partitioner
+    class Partitioner : public ::dealii::LinearAlgebra::CommunicationPatternBase
```
```CommunicationPatternBase``` was made very small not by choice but because I wasn't sure what would be common between all the vectors. If you think some of the functions in ```Partitioner``` could be useful to the others vectors just move it to the base class.

Rombur commented on include/deal.II/lac/la_parallel_vector.templates.h:
```diff
@@ -635,6 +720,47 @@ namespace parallel
+    {
+      // If no communication pattern is given, create one. Otherwise, use the
+      // given one.
+      std_cxx11::shared_ptr<const Utilities::MPI::Partitioner> comm_pattern;
```
There could also be a case where the communication pattern is not given but was already created and can be reused. I don't remember if it is what I did or not.

kronbichler replied:
In the epetra vector you store a partitioner. We should introduce that optimization here as well but for the moment my aim was to discuss the interface, keep all backward functionality (that we should restrict somehow before the release) and see if this is a sustainable path. (The optimizations also include the remainder discussed in #2535 regarding fast change between this vector and a readwrite vector.)

Rombur replied:
OK. For me, the interface is good :-) I think that deprecating as many old functions as possible (like you did) is the way to go.

kronbichler replied:
The only caveat is that with this change we need to release `ReadWriteVector` and `VectorSpaceVector` with the next release. But makes sense anyway.

There is two more thing (I will not manage to do that today), which is to teach ReadWriteVector to get data from a distributed vector and to add functions to copy from and to parallel PETSc/Trilinos vectors. These will be added but marked deprecated because we really want to promote using the ReadWriteVector. My runs revealed that I need to fix the following tests (this is a reminder to myself):
```
	790 - petsc/copy_to_dealvec.mpirun=4.debug (Failed)
	791 - petsc/copy_to_dealvec.mpirun=4.release (Failed)
	1046 - petsc/copy_to_dealvec_block.mpirun=4.debug (Failed)
	1047 - petsc/copy_to_dealvec_block.mpirun=4.release (Failed)
	6626 - mpi/parallel_block_vector_01.mpirun=1.debug (Failed)
	6627 - mpi/parallel_block_vector_01.mpirun=1.release (Failed)
	6668 - mpi/parallel_block_vector_01.mpirun=4.debug (Failed)
	6669 - mpi/parallel_block_vector_01.mpirun=4.release (Failed)
	7434 - mpi/parallel_block_vector_01.mpirun=10.debug (Failed)
	7435 - mpi/parallel_block_vector_01.mpirun=10.release (Failed)
	7632 - multigrid/step-16-04.debug (Failed)
	7633 - multigrid/step-16-04.release (Failed)
	7799 - trilinos/copy_to_dealvec.mpirun=4.debug (Failed)
	7800 - trilinos/copy_to_dealvec.mpirun=4.release (Failed)
	8143 - trilinos/copy_to_dealvec_block.mpirun=4.debug (Failed)
	8144 - trilinos/copy_to_dealvec_block.mpirun=4.release (Failed)
```
The problem in the step-16 test is that I do not have an operation `LA::distributed::Vector::add(Vector)` but only the += operation.

bangerth replied:
Just remove/deprecate the function and make people use `+=`.

Rombur commented on include/deal.II/lac/la_parallel_vector.templates.h:
```diff
@@ -635,6 +720,47 @@ namespace parallel
+
+      // fill entries from ReadWriteVector into the distributed vector,
+      // including ghost entries. this is not really efficient right now
+      // because indices are translated twice, once for
```
Missing end of the comment?

kronbichler replied:
I will append that with my next patch when I move the vector to the `ReadWriteVector::distributed` namespace (if that is what comes out). For completeness already now: The two index translations are for nth_index_in_set and once for tmp_vector::operator().

Rombur commented:
@kronbichler do you need help on this?

kronbichler commented:
I've been traveling during the last weeks so I did not get into it yet, but I will look into it towards the end of this week when I'm back.

Rombur commented:
There is no rush. I dont' want you to feel pressured to do it right now. I am just offering some help ;-)

kronbichler commented:
Now this is ready for review again. I changed the namespace to `LinearAlgebra::distributed::Vector` and added the exchange to Trilinos and PETSc vectors and `ReadWriteVector`. I did not introduce performance optimizations for re-using the communication pattern yet. Maybe @Rombur you can help me with that in a follow-up patch (or we leave it until somebody needs it)?

kronbichler commented:
There is one more item to discuss: We previously supported operations such as `is_non_negative`, `operator==` and `operator!=`. I do not think we should have them in the general vector interfaces and thus, I did not provide the functionality. This breaks backwards compatibility. I could still introduce it here and deprecate it but I don't really like that because I want the vector to also work with complex numbers and thus, I would need quite a bit of unnecessary code.

If we agree on that, I will also carry this change over to the block vector. @Rombur: I guess we want to call it `LinearAlgebra::distributed::BlockVector` or do you have other suggestions?

Finally, I will then also replace `parallel::distributed::Vector` in the test suite with `LinearAlgebra::distributed::BlockVector`.

bangerth commented:
I think it makes sense to drop the 3 functions you mention.

Rombur commented on include/deal.II/lac/la_parallel_vector.h:
```diff
@@ -0,0 +1,1434 @@
+     *
+     * Functions related to parallel functionality:
+     * <ul>
+     * <li> The function <code>compress()</code> goes through the data
```
Do we still use compress with the new interface? The other vectors don't need it anymore. Do you want to keep both interfaces or should this be deprecated?

kronbichler replied:
Right now the `compress()` and `update_ghost_values()` methods are still needed. On the one hand because they are as of now still doing the actual data exchange in the two import() methods of `ReadWriteVector` and the current vector.  I plan to address that in a later commit when I switch to lightweight conversion without copy of the locally owned range as discussed in #2535 (this includes also `operator()`, `local_element()`, and `begin()/end()`).

Rombur commented on source/lac/trilinos_epetra_communication_pattern.cc:
```diff
@@ -45,7 +45,7 @@ namespace LinearAlgebra
 
       Epetra_Map vector_space_vector_map = vector_space_vector_index_set.make_trilinos_map(*comm,
-                                           false);
+                                           true);
```
Why did you change this? VectorSpaceVector are not supposed to have overlapping dofs because this won't be allowed in Tpetra

kronbichler replied:
I merely did this because of the test `tests/trilinos/copy_to_dealvec.cc` includes a case where the trilinos vector is ghosted. I agree that it was a bad idea to change this. I will rather disallow this in the test. (Breaking a niche functionality like this straight away without a transition period with deprecated functions seems acceptable, just as we do for the `operator==`, `operator!=`, and `is_non_negative` methods.)

Rombur commented:
- I will add the optimization with the sparsity pattern once this has been merged.
 - ```LinearAlgebra::distributed::BlockVector``` sounds good.
 - Is this ready to be merged or still a work-in-progress?

kronbichler commented:
Give me an hour or two to finalize this patch by moving the parallel block vector to the same namespace as well and replace all occurrences of `parallel::distributed::Vector` in the library and testsuite by `LinearAlgebra::distributed::Vector`. Finally, I will also write a changelog.

(The next step are then the layered index set according to #2535, which will hopefully allow me to remove the access functionality of this class yet retain the overall efficiency of MatrixFree loops.)

Rombur commented on include/deal.II/lac/la_parallel_block_vector.h:
```diff
@@ -0,0 +1,599 @@
+      /**
+       * Add the vector @p V to the present one.
+       */
+      virtual BlockVector<Number> &operator+= (const VectorSpaceVector<Number> &V);
```
Does this really make sense? I am wondering if this class should really derive from ```VectorSpaceVector```.

Rombur replied:
Never mind. After seeing the implementation, I guess it makes.

kronbichler replied:
I agree it is a bit confusing to have both the global block vector as well as each individual block be `VectorSpaceVector` types. But it makes sense in my opinion for the case where I want to build block preconditioners for the purpose of solving. I want to have both a global inner product, vector addition, and so on, inside an iterative solver, i.e., I need to satisfy the `VectorSpaceVector` interface (even though I'm not sure if we ever will implement the import function; I currently throw an exception there). At the same time, access to blocks individually is needed when multiplying with block matrices or preconditioning where I might apply an inner solver, which in turn needs a vector from `VectorSpaceVector`.

Rombur replied:
Yes, I agree. I will do the same for Trilinos Vector.

Rombur commented on include/deal.II/lac/la_parallel_block_vector.templates.h:
```diff
@@ -0,0 +1,770 @@
+    BlockVector<Number> &
+    BlockVector<Number>::operator = (const value_type s)
+    {
+
```
Extra blank line

Rombur commented on include/deal.II/lac/la_parallel_block_vector.templates.h:
```diff
@@ -0,0 +1,770 @@
+      // start all requests for all blocks before finishing the transfers as
+      // this saves repeated synchronizations
+      for (unsigned int block=0; block<this->n_blocks(); ++block)
+        this->block(block).compress_start(block*10 + 8273, operation);
```
Can you add a comment for the 8273.

Rombur commented on include/deal.II/lac/la_parallel_block_vector.templates.h:
```diff
@@ -0,0 +1,770 @@
+    BlockVector<Number>::update_ghost_values () const
+    {
+      for (unsigned int block=0; block<this->n_blocks(); ++block)
+        this->block(block).update_ghost_values_start(block*10 + 9923);
```
Same thing.

Rombur commented:
Looks good :+1:

kronbichler commented:
The current version passes the test suite (except for the usual failures).

shakirbsm commented:
With these changes I am not able to build dealii as it breaks. In my case I am not using mpi and p4est or trillinos.

[detailedlog.txt](https://github.com/dealii/dealii/files/316642/detailedlog.txt)
[error.txt](https://github.com/dealii/dealii/files/316645/error.txt)

kronbichler commented:
@shakirbsm I'm on it. I just need to check that it doesn't break anything else.

Rombur commented:
@shakirbsm this should be fixed now. Thanks for the report!!

shakirbsm commented:
@Rombur it still breaks as iomanip header is missing in la_parallel_vector.h.

shakirbsm commented:
@kronbichler @Rombur It even breaks after adding that header and the error is in file
[error1.txt](https://github.com/dealii/dealii/files/318484/error1.txt)

kronbichler commented:
@shakirbsm Thanks for the info. It took me a while to reproduce on my systems because I somehow missed the violating configuration. I have a patch.

kronbichler commented:
See #2689.
